Refactor: LaTeX processing in external Docker container

b0ink opened the pull request:
**Depends on https://github.com/thoth-tech/doubtfire-deploy/pull/13**
Feasibility document: https://github.com/thoth-tech/documentation/pull/534

# Description

This update moves LaTeX processing into a dedicated TeX Live container. Instead of running `lualatex` from the [latex_config](https://github.com/thoth-tech/doubtfire-api/pull/38/files#diff-167b948375f6e2b72c3cc40acc447e3403e40953443451e7bd1501fd646a9309), the [latex_run.sh](https://github.com/b0ink/doubtfire-api/blob/ce988b26de5b18a919ff980f33cf08aff584bbeb/lib/shell/latex_run.sh) script is executed after being copied into the working directory (along with other input files such as `input.tex` and `jupynotex.py`).

`latex_run.sh` then runs `sudo docker exec ...` into the TeX Live container and executes [latex_build.sh](https://github.com/b0ink/doubtfire-api/blob/ce988b26de5b18a919ff980f33cf08aff584bbeb/lib/shell/latex_build.sh), which is in charge of running the `lualatex` commands.

`latex_run.sh` relies on two environment variables to be set:
```env
LATEX_CONTAINER_NAME=formatif-texlive-container # Name of the TeX Live container
LATEX_BUILD_PATH=/texlive/shell/latex_build.sh # Path on the TeX Live container
```

The `render_to_string()` method that renders the PDFs is now wrapped in a `generate_pdf()` method defined in [latex_helper.rb](https://github.com/b0ink/doubtfire-api/blob/ce988b26de5b18a919ff980f33cf08aff584bbeb/app/helpers/latex_helper.rb), ensuring the environment variables are set before executing `latex_run.sh` via `render_to_string`.

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- Use the modified docker-compose from https://github.com/thoth-tech/doubtfire-deploy/pull/13
- If you encounter issues with the setup, remove any previous Docker devcontainer builds and volumes.
- Run `rake test` in `doubtfire-api` to validate PDF generation.
- Optionally, run `rake db:simulate_signoff` which will create several tasks and render them with lualatex.
- You can also manually create a portfolio when logged in as a student, upload different test files, click "Create my portfolio", then run `doubtfire-api/lib/shell/generate_pdfs.sh`

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation if appropriate
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have created or extended unit tests to address my new additions
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

## TODO:
- [ ] Remove previous `texlive-install.sh` and `texlive.profile` scripts
- [x] Add TeX Live service to docker-compose in doubtfire-api

b0ink committed:
refactor: move latex processing into its own container

b0ink committed:
refactor: task pdf generation

b0ink committed:
refactor: cleanup latex docker command and compile script

b0ink committed:
fix: set correct exit status

b0ink committed:
refactor: create a new working directory for each unique latex render

b0ink committed:
chore: fix .ipynb latex math tests

b0ink committed:
chore: simplify check

b0ink committed:
refactor: move pdf generation into latex helper module

b0ink committed:
refactor: create additional work subfolder

* solves current permission issues that prevented the working folder from being rm -rf'd from the api
* this ensures that input.pdf and input.log still exist when the script successfully completes

b0ink committed:
chore: remove work directory on completion

b0ink committed:
refactor: raise error on failed pdf generation

b0ink committed:
chore: comment

b0ink committed:
chore: use latex env variables

b0ink committed:
refactor: reorganise latex script

b0ink committed:
chore: remove texlive shell folder

b0ink committed:
chore: rename latex build script

b0ink committed:
chore: use english library to replace global vars

b0ink committed:
ci: add texlive dockerfile

b0ink committed:
refactor: implement texlive docker workflow back into rails-latex

b0ink committed:
fix: set dockerfile context

JackSCarroll commented:
This looks good and latex works as intended.
Old function calls have been replaced with new functions correctly.
I like that you've added the env var checks in latex_helper.rb.

b0ink committed:
Merge remote-tracking branch 'doubtfirelms/new/scorm' into refactor/latex-docker-downsize-api

b0ink committed:
fix: allow docker exec to work in gh actions

b0ink committed:
ci: add new texlive container to gh actions

b0ink committed:
fix: ensure log is still copied if pdf gen fails

b0ink committed:
chore: remove texlive install scripts

b0ink committed:
ci: manually stop texlive service after unit tests complete

b0ink committed:
ci: add unique image scopes to fix cache conflicts

b0ink committed:
refactor: keep containers alive with sleep infinity

b0ink committed:
chore: remove unused packages

b0ink committed:
chore: replace double quotes with single quotes

b0ink committed:
refactor: keep containers alive with sleep infinity

macite committed:
feat: add ability to get individual overseer images

macite committed:
test: add overseer image api tests

macite committed:
chore(release): 8.0.41

macite committed:
fix: ensure submission history uses archive and moves

- use archive folder when unit archived
- move submission history to archive folder when unit archived
- move submission history on task abbr change
- move submission history on username change
- delete submission history on task delete

macite committed:
fix: remove archive old unit from schedule

- action is too slow for use in sidekiq
- removed from scheduled actions

macite committed:
chore(release): 8.0.42

macite committed:
fix: ensure new rollover units are not marked archived

macite committed:
feat: add ability to set an institution logo to add to header

macite committed:
Merge branch '8.0.x' of https://github.com/macite/doubtfire-api into 8.0.x

macite committed:
Merge branch '8.0.x' of https://github.com/macite/doubtfire-api into 8.0.x

macite committed:
chore: fix enum format

macite committed:
chore: update to rails 7.2

macite committed:
test: attempt to fix occasional webcal test failure

macite committed:
chore: update to rails 8 and ruby 3.4

macite committed:
chore: update ruby and bundler version in containers

macite committed:
chore: update docker install in containers

macite committed:
chore: fix minor cops

macite committed:
Merge branch 'chore/update-gems' into refactor/latex-docker-downsize-api
